File: apps/backend-dotnet/PawPals.Api/Services/DiscoverService.cs
using Microsoft.EntityFrameworkCore;
using PawPals.Api.Data;
using PawPals.Api.Models;

namespace PawPals.Api.Services;

public class DiscoverService(AppDbContext db)
{
    private static double HaversineM(double lat1, double lon1, double lat2, double lon2)
    {
        const double R = 6371000;
        var dLat = (lat2 - lat1) * Math.PI / 180;
        var dLon = (lon2 - lon1) * Math.PI / 180;
        var a = Math.Sin(dLat / 2) * Math.Sin(dLat / 2) +
                Math.Cos(lat1 * Math.PI / 180) * Math.Cos(lat2 * Math.PI / 180) *
                Math.Sin(dLon / 2) * Math.Sin(dLon / 2);
        return R * 2 * Math.Atan2(Math.Sqrt(a), Math.Sqrt(1 - a));
    }

    public async Task UpdateLocation(string userId, double lat, double lng)
    {
        await db.Users.Where(u => u.Id == userId)
            .ExecuteUpdateAsync(s => s
                .SetProperty(u => u.Latitude, lat)
                .SetProperty(u => u.Longitude, lng));
    }

    public async Task<List<object>> FindNearby(
        string userId, double radius = 50, int page = 1, int limit = 20,
        string[]? genderFilter = null, string? roleFilter = null)
    {
        var me = await db.Users.Where(u => u.Id == userId)
            .Select(u => new { u.Latitude, u.Longitude })
            .FirstOrDefaultAsync();

        if (me?.Latitude == null || me.Longitude == null) return [];

        var passExpiry = TimeSpan.FromMinutes(
            int.TryParse(Environment.GetEnvironmentVariable("PASS_EXPIRY_MINUTES"), out var m) ? m : 1);
        var expiryCutoff = DateTime.UtcNow - passExpiry;

        var swipedIds = await db.Swipes
            .Where(s => s.SwiperId == userId &&
                (s.Direction != SwipeDirection.PASS || s.CreatedAt > expiryCutoff))
            .Select(s => s.TargetId)
            .ToListAsync();

        var query = db.Users
            .Where(u => u.Id != userId && !swipedIds.Contains(u.Id)
                && u.Latitude != null && u.Longitude != null);

        if (genderFilter?.Length > 0)
        {
            var genders = genderFilter.Select(g => Enum.Parse<Gender>(g)).ToList();
            query = query.Where(u => u.Gender != null && genders.Contains(u.Gender.Value));
        }
        if (roleFilter != null && Enum.TryParse<UserRole>(roleFilter, out var role))
            query = query.Where(u => u.Role == role);

        var candidates = await query.Select(u => new
        {
            u.Id, u.Email, u.Role, u.Gender, u.DisplayName, u.Bio,
            u.Interests, u.Education, u.Zodiac, u.JobTitle, u.Company,
            u.School, u.City, u.Height, u.AvatarUrl, u.Latitude, u.Longitude,
            u.Pets
        }).ToListAsync();

        var radiusM = radius * 1000;
        return candidates
            .Select(u => new { user = u, distanceM = HaversineM(me.Latitude!.Value, me.Longitude!.Value, u.Latitude!.Value, u.Longitude!.Value) })
            .Where(x => x.distanceM <= radiusM)
            .OrderBy(x => x.distanceM)
            .Skip((page - 1) * limit).Take(limit)
            .Select(x => (object)new
            {
                user = new { x.user.Id, x.user.Email, x.user.Role, x.user.Gender, x.user.DisplayName, x.user.Bio, x.user.Interests, x.user.Education, x.user.Zodiac, x.user.JobTitle, x.user.Company, x.user.School, x.user.City, x.user.Height, x.user.AvatarUrl, x.user.Pets },
                x.distanceM
            }).ToList();
    }
}

File: apps/backend/src/discover/discover.service.ts
import { Injectable } from '@nestjs/common';
import { Prisma } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';
import { SafetyService } from '../safety/safety.service';

const USER_CARD_SELECT = {
  id: true, email: true, role: true, gender: true, displayName: true,
  bio: true, interests: true, education: true, zodiac: true,
  jobTitle: true, company: true, school: true, city: true,
  height: true, avatarUrl: true,
};

function haversineM(lat1: number, lon1: number, lat2: number, lon2: number): number {
  const R = 6371000;
  const dLat = (lat2 - lat1) * Math.PI / 180;
  const dLon = (lon2 - lon1) * Math.PI / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(lat1 * Math.PI / 180) * Math.cos(lat2 * Math.PI / 180) * Math.sin(dLon / 2) ** 2;
  return R * 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
}

@Injectable()
export class DiscoverService {
  constructor(
    private prisma: PrismaService,
    private safetyService: SafetyService,
  ) {}

  async updateLocation(userId: string, lat: number, lng: number) {
    await this.prisma.user.update({
      where: { id: userId },
      data: { latitude: lat, longitude: lng },
    });
    return { ok: true };
  }

  async findNearby(
    userId: string,
    radius = 5,
    page = 1,
    limit = 20,
    genderFilter: string[] = [],
    roleFilter: string | null = null,
  ) {
    const me = await this.prisma.user.findUnique({
      where: { id: userId },
      select: { latitude: true, longitude: true },
    });
    if (!me?.latitude || !me?.longitude) return [];

    const passExpiryMinutes = parseInt(process.env.PASS_EXPIRY_MINUTES ?? '1');
    const passExpiryDate = new Date(Date.now() - passExpiryMinutes * 60 * 1000);

    const swipes = await this.prisma.swipe.findMany({
      where: {
        swiperId: userId,
        OR: [
          { direction: { not: 'PASS' as any } },
          { direction: 'PASS' as any, createdAt: { gt: passExpiryDate } },
        ],
      },
      select: { targetId: true },
    });
    const swipedIds = swipes.map((s) => s.targetId);
    const blockedIds = await this.safetyService.getBlockedUserIds(userId);

    const where: Prisma.UserWhereInput = {
      id: { not: userId, notIn: [...swipedIds, ...blockedIds] },
      latitude: { not: null },
      longitude: { not: null },
      ...(genderFilter.length > 0 && genderFilter.length < 3
        ? { gender: { in: genderFilter as any } }
        : {}),
      ...(roleFilter !== null ? { role: roleFilter as any } : {}),
    };

    const candidates = await this.prisma.user.findMany({
      where,
      select: { ...USER_CARD_SELECT, latitude: true, longitude: true },
    });

    const radiusM = radius * 1000;
    const withDistance = candidates
      .map((u) => ({
        user: u,
        distanceM: haversineM(me.latitude!, me.longitude!, u.latitude!, u.longitude!),
      }))
      .filter((u) => u.distanceM <= radiusM)
      .sort((a, b) => a.distanceM - b.distanceM);

    const offset = (page - 1) * limit;
    return withDistance.slice(offset, offset + limit).map(({ user, distanceM }) => {
      const { latitude, longitude, ...rest } = user;
      return { user: rest, distanceM };
    });
  }
}
